blosc/shuffle-altivec.c

```diff
@@ -15,61 +16,145 @@
 shuffle2_altivec(uint8_t* const dest, const uint8_t* const src,
-                 const int32_t vectorizable_elements, const int32_t total_elements) {
-  static const int32_t bytesoftype = 2;
+                 const int32_t vectorizable_elements, const int32_t total_elements) 
+{
+  static const __vector uint8_t even = (const __vector uint8_t) {0x00, 0x02, 0x04, 0x06, 0x08, 0x0a, 0x0c, 0x0e,
+                                                                 0x10, 0x12, 0x14, 0x16, 0x18, 0x1a, 0x1c, 0x1e};
+  static const __vector uint8_t odd = (const __vector uint8_t) {0x01, 0x03, 0x05, 0x07, 0x09, 0x0b, 0x0d, 0x0f,
+                                                                0x11, 0x13, 0x15, 0x17, 0x19, 0x1b, 0x1d, 0x1f};
+  const int32_t offset = total_elements%16;
+  static __vector uint8_t permute_lo, permute_hi;
+  const int32_t bytesoftype = 2;
   int32_t j;
-  int k;
-  uint8_t* dest_for_jth_element;
-  __m128i xmm0[2], xmm1[2];
-
-  for (j = 0; j < vectorizable_elements; j += sizeof(__m128i)) {
-    /* Fetch 16 elements (32 bytes) then transpose bytes, words and double words. */
-    for (k = 0; k < 2; k++) {
-      xmm0[k] = _mm_loadu_si128((__m128i*)(src + (j * bytesoftype) + (k * sizeof(__m128i))));
-      xmm0[k] = _mm_shufflelo_epi16(xmm0[k], 0xd8);
-      xmm0[k] = _mm_shufflehi_epi16(xmm0[k], 0xd8);
-      xmm0[k] = _mm_shuffle_epi32(xmm0[k], 0xd8);
-      xmm1[k] = _mm_shuffle_epi32(xmm0[k], 0x4e);
-      xmm0[k] = _mm_unpacklo_epi8(xmm0[k], xmm1[k]);
-      xmm0[k] = _mm_shuffle_epi32(xmm0[k], 0xd8);
-      xmm1[k] = _mm_shuffle_epi32(xmm0[k], 0x4e);
-      xmm0[k] = _mm_unpacklo_epi16(xmm0[k], xmm1[k]);
-      xmm0[k] = _mm_shuffle_epi32(xmm0[k], 0xd8);
-    }
-    /* Transpose quad words */
-    for (k = 0; k < 1; k++) {
-      xmm1[k * 2] = _mm_unpacklo_epi64(xmm0[k], xmm0[k + 1]);
-      xmm1[k * 2 + 1] = _mm_unpackhi_epi64(xmm0[k], xmm0[k + 1]);
-    }
+  __vector uint8_t inp0, inp1, out0, out1, low;
+  uint8_t* const dest1 = &dest[total_elements];
+  
+  //printf("vectorizable_elements: %d total_elements: %d dst: %d dst1: %d delta: %d offset: %d\n", vectorizable_elements, total_elements, dest, dest1, dest1-dest, offset);
+  
+  if (offset)
+  {
+     // Initialize the two permutation vectors
+     permute_lo = gen_permute_low(offset);
+     permute_hi = gen_permute_high(offset);
+     low = (__vector uint8_t) {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+  }
+  for (j = 0; j < vectorizable_elements; j += 16) 
+  {
+    /* Fetch 16 elements (32 bytes) */
+    inp0 = vec_ld(bytesoftype*j, src);
+    inp1 = vec_ld(bytesoftype*j+16, src);
+    /* Transpose vectors */
+    out0 = vec_perm(inp0, inp1, even);
+    out1 = vec_perm(inp0, inp1, odd);
     /* Store the result vectors */
-    dest_for_jth_element = dest + j;
-    for (k = 0; k < 2; k++) {
-      _mm_storeu_si128((__m128i*)(dest_for_jth_element + (k * total_elements)), xmm1[k]);
+    vec_st(out0, j, dest);
+    if (offset)
+    {// misaligned store
```

Comment:
The opening curly brace should go up.  OTOH, the closing brace is recommended to be in their own line.